app/controllers/admin/users_controller.rb

```diff
@@ -191,25 +190,31 @@ def revoke_admin
   end
 
   def grant_admin
-    guardian.ensure_can_grant_admin!(@user)
-    if current_user.has_any_second_factor_methods_enabled?
-      second_factor_authentication_result = current_user.authenticate_second_factor(params, secure_session)
-      if second_factor_authentication_result.ok
-        @user.grant_admin!
-        StaffActionLogger.new(current_user).log_grant_admin(@user)
-        render json: success_json
-      else
-        failure_payload = second_factor_authentication_result.to_h
-        if current_user.security_keys_enabled?
-          Webauthn.stage_challenge(current_user, secure_session)
-          failure_payload.merge!(Webauthn.allowed_credentials(current_user, secure_session))
-        end
-        render json: failed_json.merge(failure_payload)
-      end
-    else
-      AdminConfirmation.new(@user, current_user).create_confirmation
+    manager = SecondFactorVerificationManager.new(current_user)
+    manager.on_no_second_factors_enabled do
```

Comment:
this block doing a render is a bit odd, can we avoid it being responsible for render?